File: jinding_config_build/tmp_V12/state_machine.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include "cJSON.h"

#define DISABLE "disable"
#define TRAGER  "trager"
#define ON      "on"
#define OFF     "off"

#define MAJOR_PATH  "module/major.rule"
#define INDEP_PATH  "module/indep.conf"
#define SCENE_PATH  "module/scene.conf"
#define DEPEND_PATH "module/depends.conf"
#define STATE_PATH  "module/state.json"

typedef enum {
    NIL = 0,
    EN,
    DIS,
} STATE;

cJSON *parse_from_file(char *path)
{
    char buff[1024] = "";
    char *total = NULL;
    int nread = 0;
    FILE *file = fopen(path, "r");
    if(file) {
        while((nread = fread(buff, 1, 1023, file)) > 0) {
            if(total == NULL) {
                total = strdup(buff);
            } else {
                total = realloc(total, nread + strlen(total) + 1);
                strcat(total, buff);
            }
        }
        fclose(file);
    }
    cJSON *ret = cJSON_Parse(total);
    free(total);
    return ret;
}

cJSON *get_major()
{
    return parse_from_file(MAJOR_PATH);
}

cJSON *get_indep()
{
    return parse_from_file(INDEP_PATH);
}

cJSON *get_scene()
{
    return parse_from_file(SCENE_PATH);
}

cJSON *get_depend()
{
    return parse_from_file(DEPEND_PATH);
}

cJSON *get_state()
{
    return parse_from_file(STATE_PATH);
}

bool check_json_tab(cJSON *tar, char *name)
{
    if(cJSON_GetObjectItem(tar, name)) {
        return true;
    } else {
        return false;
    }
}

bool check_json_list(cJSON *tar, char *name)
{
    bool ret = false;
    int ntar = cJSON_GetArraySize(tar);
    int i;
    for(i = 0; i < ntar; i ++) {
        cJSON *item = cJSON_GetArrayItem(tar, i);
        if(strcmp(item->valuestring, name) == 0) {
            ret = true;
        }
    }
    return ret;
}

bool check_scene(char *action)
{
    cJSON *scene = get_scene();
    bool ret = check_json_tab(scene, action);
    cJSON_Delete(scene);
    return ret;
}

bool check_depend(char *action)
{
    cJSON *depend = get_depend();
    bool ret = check_json_list(depend, action);
    cJSON_Delete(depend);
    return ret;
}

bool check_indep(char *action)
{
    cJSON *indep = get_indep();
    bool ret = check_json_list(indep, action);
    cJSON_Delete(indep);
    return ret;
}

bool check_major(char *action)
{
    cJSON *major = get_major();
    bool ret = check_json_tab(major, action);
    cJSON_Delete(major);
    return ret;
}

bool trager_json_state(cJSON *state, char *action)
{
    cJSON *item = cJSON_GetObjectItem(state, action);
    if(item->type == cJSON_False) {
        return true;
    } else {
        return false;
    }
}

cJSON *set_state(cJSON *state, char *name, STATE action)
{
    switch(action) {
        case NIL: {
            cJSON_ReplaceItemInObject(state, name, cJSON_CreateBool(trager_json_state(state, name)));
            break;
        }
        case EN: {
            cJSON_ReplaceItemInObject(state, name, cJSON_CreateTrue());
            break;
        }
        case DIS: {
            cJSON_ReplaceItemInObject(state, name, cJSON_CreateFalse());
            break;
        }
        default: break;
    } 
    return state;
}

cJSON *delete_item_from(cJSON *array, char *name)
{
    int narray = cJSON_GetArraySize(array);
    int i;
    for(i = 0; i < narray; i++) {
        cJSON *item = cJSON_GetArrayItem(array, i);
        // printf("delete && find %s\n", item->valuestring);
        if(strcmp(item->valuestring, name) == 0) {
            cJSON_DeleteItemFromArray(array, i);
            break;
        }
    }
    return array;
}

cJSON *get_disable_deps(cJSON *state, cJSON *major)
{
    cJSON *deps = get_depend();
    // char *out = cJSON_Print(deps);
    // printf("deps before %s\n", out);
    // free(out);

    int nmajor = cJSON_GetArraySize(major);
    int i;
    for(i = 0; i < nmajor; i++) {
        cJSON *imajor = cJSON_GetArrayItem(major, i);
        cJSON *major_state = cJSON_GetObjectItem(state, imajor->string);
        if(major_state && major_state->type == cJSON_True) {
            // printf("%s is true\n", major_state->string);
            cJSON *mdeps = cJSON_GetObjectItem(imajor, "deps");
            int ndeps = cJSON_GetArraySize(mdeps);
            int j = 0;
            for(j = 0; j < ndeps; j++) {
                cJSON *idep = cJSON_GetArrayItem(mdeps, j);
                // printf("%s %s ", imajor->string, idep->valuestring);
                if(cJSON_GetObjectItem(deps, idep->string)) {
                    delete_item_from(deps, idep->valuestring);
                    // cJSON_DeleteItemFromObject(deps, idep->valuestring);
                    // printf("delete %s\n", idep->valuestring);
                }
            }
        }
    }
    // out = cJSON_Print(deps);
    // printf("deps %s\n", out);
    // free(out);
    return deps;
}

cJSON *update_deps(cJSON *state, cJSON *disables)
{
    cJSON *deps = get_depend();
    int ndeps = cJSON_GetArraySize(deps);
    int i;
    for(i = 0; i < ndeps; i++) {
        cJSON *dis = cJSON_GetArrayItem(deps, i);
        if(check_json_list(disables, dis->valuestring)) {
            cJSON_ReplaceItemInObject(state, dis->valuestring, cJSON_CreateString(DISABLE));
        } else {
            cJSON *st = cJSON_GetObjectItem(state, dis->valuestring);
            if(st->type == cJSON_String) {
                cJSON_ReplaceItemInObject(state, dis->valuestring, cJSON_CreateFalse());
            }
        }
    }
    cJSON_Delete(deps);
    return state;
}

cJSON *major_rule(cJSON *state, char *name, STATE action)
{
    set_state(state, name, action);
    // char *out = cJSON_Print(state);
    // printf("state is %s\n", out);
    // free(out);
    cJSON *major = get_major();
    cJSON *rela = cJSON_GetObjectItem(major, name);
    int size = cJSON_GetArraySize(rela);
    // printf("size of rela is %d\n", size);
    int i = 0;
    cJSON *deps = NULL;
    for(i; i < size; i++) {
        cJSON *item = cJSON_GetArrayItem(rela, i);
        // printf("%s -> %s\n", item->string, item->valuestring);
        if(item->type == cJSON_String && strcmp(item->valuestring, "exclusive") == 0) {
            cJSON_ReplaceItemInObject(state, item->string, cJSON_CreateFalse());
        }
    }
    cJSON *disables = get_disable_deps(state, major);
    update_deps(state, disables);
    cJSON_Delete(disables);
    cJSON_Delete(major);
    return state;
}

cJSON *Fun(cJSON *state, char *name, STATE action)
{
    if(check_depend(name)) {
        set_state(state, name, NIL);
    } else if(check_indep(name)) {
        set_state(state, name, action);
    } else if(check_major(name)) {
        major_rule(state, name, action);
    }
    return state;
}

cJSON *scene_check(cJSON *state)
{
    cJSON *sce = get_scene();
    int nsce = cJSON_GetArraySize(sce);
    int i;
    for(i = 0; i < nsce; i++) {
        cJSON *sitem = cJSON_GetArrayItem(sce, i);
        if(cJSON_GetObjectItem(state, sitem->string)->type == cJSON_True) {
            int nsitem = cJSON_GetArraySize(sitem);
            int j = 0;
            for(j = 0; j < nsitem; j++) {
                cJSON *action_item = cJSON_GetArrayItem(sitem, j);
                if(action_item->type == cJSON_True &
                cJSON_GetObjectItem(state, action_item->string)->type != cJSON_True) {
                    cJSON_ReplaceItemInObject(state, sitem->string, cJSON_CreateFalse());
                    break;
                }
            }
        }
    }
    return state;
}

cJSON *batch_fun(cJSON *state, char *sce_name, STATE action)
{
    set_state(state, sce_name, action);
    cJSON *sce = get_scene();
    cJSON *sce_item = cJSON_GetObjectItem(sce, sce_name);
    int nitem = cJSON_GetArraySize(sce_item);
    int i;
    for(i = 0; i < nitem; i++) {
        cJSON *st = cJSON_GetArrayItem(sce_item, i);
        if(st->type == cJSON_True) {
            Fun(state, st->string, action);
        }
    }
    cJSON_Delete(sce);
    return state;
}

int main(int argc, char const *argv[])
{
    /* code */
    cJSON *state = get_state();
    if (state == NULL) {
        perror("get state failed\n");
    }
    if(argc == 1) {

    } else {
        cJSON *item = cJSON_GetObjectItem(state, argv[1]);
        if(item->type == cJSON_String && (strcmp(item->valuestring, DISABLE) == 0)) {
            
        } else {
            if(check_scene(argv[1])) {
                //scene~
                cJSON *sce_state = cJSON_GetObjectItem(state, argv[1]);
                if(sce_state->type == cJSON_False) {
                    cJSON *sce = get_scene();
                    int nsce = cJSON_GetArraySize(sce);
                    int i = 0;
                    for(i = 0; i < nsce; i++) {
                        cJSON *s = cJSON_GetArrayItem(sce, i);
                        cJSON *st = cJSON_GetObjectItem(state, s->string);
                        if(st->type == cJSON_True) {
                            batch_fun(state, s->string, false);
                        }
                    }
                    batch_fun(state, argv[1], EN);
                } else {
                    batch_fun(state, argv[1], DIS);
                }
            } else {
                Fun(state, argv[1], NIL);
                scene_check(state);
                // not a scene
            }
        }
    }
    char *out = cJSON_PrintUnformatted(state);
    printf("%s\n", out);
    FILE *file = fopen(STATE_PATH, "w");
    fwrite(out, 1, strlen(out), file);
    fclose(file);

    free(out);
    cJSON_Delete(state);    
    return 0;
}

File: jinding_config_build/tmp_V12/state_machine.lua
#!/usr/bin/lua
-- local json = require("cjson")
local json = require("cjson.safe")
sm = {}

local zone = -1
function dump_tab( tab )
    zone = zone + 1
    for k, v in pairs(tab) do
        local str = "\t"
        str = str:rep(zone)
        if type(v) == 'table' then
            print(str..k)
            dump_tab(v)
        else
            print(str..k, v)
        end
    end
    zone = zone - 1
end

function tab_update(tar, sub)
    if type(tar) == 'table' and type(sub) == 'table' then
        for ks, vs in pairs(sub) do
            tar[ks] = vs
        end
    end
    return tar
end

function list_extend(tar, sub)
    if type(tar) == 'table' and type(sub) == 'table' then
        for k, v in ipairs(sub) do
            table.insert(tar, v)
        end
    end
end

local _menu_ = 'module/'
local _major_path_ = _menu_..'major.rule'
local _indep_path_ = _menu_..'indep.conf'
local _scene_path_ = _menu_..'scene.conf'
local _state_path_ = _menu_..'state.json'
local _depend_path_ = _menu_..'depends.conf'
local _reset_path_ = _menu_..'reset.conf'
local DISABLE = 'disable'
local TRAGER = 'trager'
local ON = 'on'
local OFF = 'off'

local f = io.open(_major_path_)
local major = json.decode(f:read())
f:close()
f = io.open(_depend_path_)
local dep = json.decode(f:read())
f:close()
f = io.open(_indep_path_)
local indep = json.decode(f:read())
f:close()
f = io.open(_scene_path_)
local sce = json.decode(f:read())
f:close()
f = io.open(_reset_path_)
local reset = json.decode(f:read())
f:close()

local function _check_list_(tar_name, name)
    for n, v in ipairs(name) do
        if v == tar_name then
            return true
        end
    end
    return false
end

local function _check_tab_(tar_name, tab)
    for n, name in pairs(tab) do
        if (type(n) == 'string' and n == tar_name) or 
            (type(name) == 'string' and name == tar_name) then
            return true
        end
    end
    return false
end

local function check_indep(tar_name)
    return _check_tab_(tar_name, indep)
end

local function check_dep(tar_name)
    return _check_tab_(tar_name, dep)
end 

local function check_major(tar_name)
    return _check_tab_(tar_name, major)
end

local function check_scene(tar_name)
    return _check_tab_(tar_name, sce)
end

local function check_reset(tar_name)
    return _check_list_(tar_name,reset)
end

local function get_active_deps(state)
    
end

local function get_disable_deps(state)
    f = io.open(_depend_path_)
    local d = json.decode(f:read())
    f:close()
    tmp = {}
    -- print('check!!!!!!', json.encode(state))
    for chief, r in pairs(major) do
        if state[chief] == true then
            list_extend(tmp, r['deps'])
        end
    end
    -- print('dis !!!', json.encode(tmp))
    for n, t in pairs(tmp) do
        -- print(n, t)
        for nd, vd in pairs(d) do
            if t == vd then
                table.remove(d, nd)
                break
            end
        end
    end
    return d
end

local function deps_update(state, dis)
    f = io.open(_depend_path_)
    local dep = json.decode(f:read())
    f:close()
    for n, d in pairs(dep) do
        -- print('check', d)
        -- print("------!!!!------")
        -- dump_tab(dis)
        -- print("------!!!!------")
        if _check_list_(d, dis) then
            -- print("DISABLE", d)
            state[d] = DISABLE
        else
            -- print('Enable',d, state[d])
            if state[d] == DISABLE then
                state[d] = false
            end
            -- print(d, state[d])
        end
    end
    return state
end

local function major_rule(state, name, action)
    if action == nil then
        state[name] = not(state[name])
    else
        state[name] = action
    end
    rela = major[name]
    for chief, r in pairs(rela) do
        if state[chief] ~= nil and r == 'exclusive' then
            state[chief] = false
        end
    end
    d = get_disable_deps(state)
    state = deps_update(state, d)
    -- print('----------disable------------')
    -- dump_tab(d)
    -- print('----------disable------------')
    for nd, vd in pairs(d) do
        state[vd] = DISABLE
    end
end

local function F(state, name, action)
    -- print(name, state[name])
    if check_dep(name) then
        state[name] = not(state[name])
        -- state[name] = false
    elseif check_indep(name) then
        if action == nil then
            state[name] = not(state[name])
        else
            state[name] = action
        end
    elseif check_major(name) then
        -- print('major',name, '!!!!!!')
        major_rule(state,name,action)
        -- print(name, state[name])
    end
    return state
end

local function scene_check(state)
    for ks, vs in pairs(sce) do
        -- print(ks, vs)
        if state[ks] then
            -- print('true!!!')
            for n, v in pairs(vs) do
                -- print(n, v)
                if v and not state[n] then
                    -- print('find!!!')
                    state[ks] = false
                    break
                end
            end
        end
    end
    return state
end

local function batch_f(state, sce_name, action)
    sce_batch = sce[sce_name]
    if action == nil then
        state[sce_name] = not(state[sce_name]) 
    else
        state[sce_name] = action
    end
    for n, s in pairs(sce_batch) do
        if s then
            state = F(state, n, state[sce_name])
            -- print("!!!!!!", n)
            -- print('state', json.encode(state))

        end
    end
    return state
    
end

local function state_machine(name)
    local f = io.open(_state_path_)
    local state = json.decode(f:read())
    f:close()
    if state[name] == DISABLE then
        -- print("disable", name)
        return state
    else
        if check_reset(name) then
            -- print("reset!!!")
            for k, v in pairs(major) do
                if k ~= 'dependence' then
                    state[k] = false
                end
            end
            for n, k in ipairs(indep) do
                state[k] = false
            end
            for n, k in ipairs(dep) do
                state[k] = DISABLE
            end
            for k, v in pairs(sce) do
                state[k] = false
            end
        elseif check_scene(name) then
            if state[name] == false then
                for s, t in pairs(sce) do
                    if state[s] then
                        state = batch_f(state,s,false)
                    end
                end
            end
            batch_f(state,name,not(state[name]))
        else
            F(state, name, nil)
            scene_check(state)
            -- print("not a scene", name)
        end
    end
    -- print("---------state new--------")
    -- dump_tab(state)
    -- print("---------state new--------")
    f = io.open(_state_path_, 'w')
    s = json.encode(state)
    f:write(s)
    f:close()
    -- print(state)
    print(json.encode(state))
    return state
end

-- function sm.state_machine(name)
--     return state_machine(name)
-- end

-- state_machine('摆风')
-- return sm
-- state_machine('照明二')
-- state_machine('洗浴场景')
state_machine(arg[1])
-- print(os.clock())
-- print(arg[1])
-- state_machine('取暖')

-- a = {
--     ['a'] = 'sss',
--     ['b'] = 'ddd'
-- }
-- b = {
--     a = 'dfdf',
--     b,
--     1,
--     2,
--     3,
--     'a',
-- }
-- -- dump_tab( tab_update(a,b))
-- c = {
--     1,2,3
-- }
-- d = {
--     1,3,'sdf'
-- }
-- dump_tab(tab_update(c,d))